### DISCUSSION TITLE
Match single wildcard character with underscore ("_")

### DISCUSSION THREAD

**Discussion by tero-kiituri** [pocketbase/pocketbase#5338] [Q&A] [answered]
I am using `LIKE` expression using the `~` operator: `key ~ "foo"`. This works fine. Matching with `%`  (`key ~ "f%o"`) works fine too.

However, percent sign (`%`) matches if there is 0 to n characters and I have a use case where I should match to exactly one character. I am trying to do this with an underscore character (`_`). My query is `key ~ "f_o"`, but I just can't get it to work. The query does not return any results.

The underscore is single character wildcard in SQLite. If I interpret correctly there are even some test cases for the underscore character in `filter_test.go` and those are tests are passing.

Am I missing something? Should the underscore wildcard work?

**Comment by ganigeorgiev** [comment, marked as answer]
When no explicit `%` wildcard is specified the `~` operator will wrap the string with `%` and the filter value will be treated as a plain string (aka. the underscore in this case has no special meaning).

When there is an explicit `%` in the string then the value is treated as custom LIKE expression and it is up to the user to escape the `_` and `%` symbols from the text literals. There is no support for specifying a single character wildcard without having at least one `%` at the moment (_this auto wrapping and escape behavior is mostly kept for backward compatibility with earlier versions and may change in the future or at least use a different single wildcard character as `_` is too common in string literals and will require manual `\` escaping_).

For now if you want to match a single character between `f` and `o` you'll have to specify also the `%` wildcard character, for example: `key ~ '%f_o%'`.

**Reply by tero-kiituri** [reply]
Thanks for an excellent answer once again. A single char wildcard would be helpful, but I'll find a workaround.

My first thought is a bit hacky. I can achieve almost what I want with `key ~ '%f_o' && key ~ 'f_o%'`. That works as long as the underscore isn't at either end of a search term. If the underscore is at ether end of the search term a search may return wrong keys. Wrong keys should be longer than the search term, so I should be able to get a correct result set by filtering out those too long keys. I have to do this final filtering in a client app, but that's OK since nobody should use my server API directly.

This requires more thinking, but it could work 😄.

**Reply by tero-kiituri** [reply]
BTW it would be helpful if Pocketbase had a string length filter option. Maybe that could be a new feature request.
